framework/include/materials/Material.h

```diff
@@ -13,13 +13,39 @@
 #include "MaterialBase.h"
 #include "Coupleable.h"
 #include "MaterialPropertyInterface.h"
+#include "FEProblemBase.h"
+
+#include <string>
 
 // forward declarations
 class Material;
 
 template <>
 InputParameters validParams<Material>();
 
+/**
+ * Helper class for deferred getting of material properties after the construction
+ * phase for materials. This enables "optional material properties" in materials.
+ * It works by returning a reference to a pointer to a material property (rather
+ * than a reference to the property value). The pointer will be set to point to
+ * either an existing material property or to nullptr if the requested property
+ * does not exist.
+ */
+class OptionalMaterialPropertyProxyBase
+{
+public:
+  OptionalMaterialPropertyProxyBase(const std::string & name, Material * material)
+    : _name(name), _material(material)
+  {
+  }
+  virtual ~OptionalMaterialPropertyProxyBase() {}
+  virtual void resolve() = 0;
+
+protected:
+  const std::string _name;
+  Material * _material;
+};
```

Comment:
Don't need to store the material here; just pass `Material &` into resolve within `Material::resolveWhateverMethodThing()`

Reply:
Yeah, another benefit of the new Material local storage design.